### PR TITLE
Machine constraints

### PR DESCRIPTION
**Pull request by huwshimi**
Fixes: #1379. Fixes: #1982.

### CODE REVIEW

**Comment by huwshimi** [comment]
QA:
- start a new model
- open the machine view and add a machine with some constraints
- the constraints should be visible on the added machine
- use the context menu on the machine to open the update constraints form
- change some constraints
- the new constraints should be visible on the machine
- open the deployment flow
- scroll down to the machines section
- the machine constraints should be the same as the updated constraints

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7357/

File: jujugui/static/gui/src/app/components/machine-view/machine/machine.js [hunk 1]
```diff
@@ -71,13 +73,140 @@ YUI.add('machine-view-machine', function() {
...
+            break;
+        }
+      });
+      return {
+        arch: arch || null,
+        cpuCores: cpuCores || null,
+        cpuPower: cpu || null,
+        disk: disk || null,
+        mem: mem || null
+      };
```
Comment (Lukewh) [inline jujugui/static/gui/src/app/components/machine-view/machine/machine.js]: Might be cleaner to do something like:
```
let types = {
  'cpu-power': null,
  cores: null,
  mem: null,
  'root-disk': null,
  arch: null;
};
```
So you can remove the switch and do
```
const keyVal = part.split('=');
types[keyVal[0]] = keyVal[1];
```
Then assign to what you want in the return
```
return {
  arch: types.arch,
  cpuCores: types.cores,
  cpuPower: types['cpu-power'],
  disk: types['root-disk'],
  mem: mem
};
```
Saves a few lines.
The ideal in this situation is to use a consistent naming convention, but that's not always possible.

Comment (Lukewh) [inline jujugui/static/gui/src/app/components/machine-view/machine/machine.js]: Worth noting, I don't like switch statements unless it's for numbers.

Comment (huwshimi) [inline jujugui/static/gui/src/app/components/machine-view/machine/machine.js]: So much better, thankyou!

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7362/

File: jujugui/static/gui/src/app/components/machine-view/machine/machine.js [hunk 2]
```diff
@@ -71,13 +73,125 @@ YUI.add('machine-view-machine', function() {
...
+      this.setState({constraints: constraints});
+    },
+
+    /**
+      Set the new constraints on the machine.
+    */
+    _setConstraints: function() {
+      const constraints = this.state.constraints;
+      const series = constraints.series || null;
+      delete constraints.series;
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/machine-view/machine/machine.js]: why? Can you add a code comment plz

File: jujugui/static/gui/src/app/components/machine-view/machine/machine.js [hunk 3]
```diff
@@ -88,30 +202,47 @@ YUI.add('machine-view-machine', function() {
...
+      const machine = this.props.machine;
+      const hardware = machine.hardware || this._getConstraints() || {};
+      let hardwareDetails;
+      let cpu = hardware.cpuPower;
+      let disk = hardware.disk;
+      let mem = hardware.mem;
+      let cpuCores = hardware.cpuCores;
+      if (cpuCores || cpu || disk || mem) {
+        let details = [];
+        if (cpuCores) {
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/machine-view/machine/machine.js]: I wonder if these checks to generate this string shouldn't be moved to a util as you've duplicated it twice now.

File: jujugui/static/gui/src/app/components/machine-view/machine/machine.js [hunk 4]
```diff
@@ -71,13 +73,125 @@ YUI.add('machine-view-machine', function() {
...
+    _getConstraints: function() {
+      const constraints = this.props.machine.constraints || '';
+      let types = {
+        'cpu-power': null,
+        cores: null,
+        mem: null,
+        'root-disk': null,
+        arch: null
+      };
+      // The constraints are in the format:
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/machine-view/machine/machine.js]: Why are the constraints a string here?

File: jujugui/static/gui/src/app/components/machine-view/machine/machine.js [hunk 5]
```diff
@@ -71,13 +73,125 @@ YUI.add('machine-view-machine', function() {
...
+    },
+
+    /**
+      Get the constraints for the machine.
+
+      @returns {Object} the machine constraints.
+    */
+    _getConstraints: function() {
+      const constraints = this.props.machine.constraints || '';
+      let types = {
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/machine-view/machine/machine.js]: Do we not have a master list of these types somewhere? I imagine that would go well in jujulib

Comment (hatched) [inline jujugui/static/gui/src/app/components/machine-view/machine/machine.js]: There are probably other places in the code base that use these, maybe a follow-up to centralize?

Comment (huwshimi) [inline jujugui/static/gui/src/app/components/machine-view/machine/machine.js]: You're right we have a list in api.js. I'll update this to use that.

File: jujugui/static/gui/src/app/components/machine-view/machine/machine.js (hunk 4 above)
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/machine-view/machine/machine.js]: The machine stores the constraints as a string. I've updated the comment to clarify that.

As to why that's the case, I'm not entirely sure.

File: jujugui/static/gui/src/app/components/machine-view/machine/machine.js (hunk 3 above)
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/machine-view/machine/machine.js]: Done.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7370/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7371/

**Comment by huwshimi** [comment]
Thanks @Lukewh and @hatched :shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**